ekan0ra/bot.py

```diff
@@ -0,0 +1,448 @@
+                else:
+                    self.say(self.channel, 'No one is in queue.')
+
+            elif msg == '.masters':
+                self.say(
+                    self.channel,
+                    'My current masters are: %s' %
+                        ', '.join(self.channel_admins_list)
+                )
+
+            elif msg.startswith('.add'):
+                try:
+                    nick = msg.split()[1]
+                    if nick in self.channel_admins_list:
+                        self.say(
+                            self.channel, '%s is already a master.' % nick)
+                        application_logger.info('%s is already an admin.',
+                            nick)
+                    else:
+                        self.add_admin(nick)
+                        self.say(self.channel,'%s is a master now.' % nick)
+                        application_logger.info('%s became an admin.', nick)
+                except Exception, err:
+                    application_logger.error(
+                        'Error adding admin!', exc_info=True
+                    )
+
+            elif msg.startswith('.rm'):
+                try:
+                    nick = msg.split()[1]
+                    self.remove_admin(nick)
+                    self.say(self.channel, '%s removed from admin.' % nick)
+                    application_logger.info('%s removed from admin.', nick)
+                except Exception, err:
+                    application_logger.error(
+                        'Error removing admin!', exc_info=True
+                    )
+
+            elif msg.lower().startswith('.pingall'):
+                self.pingmsg = msg.lower().lstrip('.pingall')
+                self.names(channel).addCallback(self.pingall)
+        # end processing admin message
+
+
+        # User wants to ask a question
+        if msg == '!'  and self.islogging:
+            self.qs_queue.enqueue(user.nick)
+            self.describe(
+                self.channel, 'Queue: [%r]' % ', '.join(self.qs_queue))
+
+        # User no longer wants to ask a question; remove them from queue
+        elif msg in ('!-', '!!')  and self.islogging:
+            self.qs_queue.dequeue(user.nick)
+            self.describe(
+                self.channel, 'Queue: [%r]' % ', '.join(self.qs_queue))
+
+        elif msg in ('!', '!-', '!!') and not self.islogging:
+            self.say(
+                self.channel,
+                '%s: No session is going on, feel free to ask a question. '
+                    'You do not have to type %s' % (user.nick, msg)
+            )
+        # end processing question indicator   
+
+        elif msg == '.givemelogs':
+            if not self.last_log_filename:
+                self.say(user.nick, 'Sorry, I do not have the last log.')
+                application_logger.warning('Could not find last class log!')
+            else:
+                sys.argv = ['fpaste', self.last_log_filename]
```

Comment:
One thing that I have in mind for sometime is that if the class is over then maybe we can store the fpaste url. Also I had plans that a nick can only request twice during a session.